crates/accelerate/src/circuit_library/pauli_evolution.rs

```diff
@@ -0,0 +1,330 @@
+    sparse_paulis: &Bound<PyList>,
+    insert_barriers: bool,
+    do_fountain: bool,
+) -> PyResult<CircuitData> {
+    let py = sparse_paulis.py();
+    let num_paulis = sparse_paulis.len();
+    let mut paulis: Vec<String> = Vec::with_capacity(num_paulis);
+    let mut indices: Vec<Vec<u32>> = Vec::with_capacity(num_paulis);
+    let mut times: Vec<Param> = Vec::with_capacity(num_paulis);
+    let mut global_phase = Param::Float(0.0);
+    let mut modified_phase = false; // keep track of whether we modified the phase
+
+    for el in sparse_paulis.iter() {
+        let tuple = el.downcast::<PyTuple>()?;
+        let pauli = tuple.get_item(0)?.downcast::<PyString>()?.to_string();
+        let time = Param::extract_no_coerce(&tuple.get_item(2)?)?;
+
+        if pauli.as_str().chars().all(|p| p == 'i') {
+            global_phase = radd_param(global_phase, time, py);
+            modified_phase = true;
+            continue;
+        }
+
+        paulis.push(pauli);
+        times.push(time); // note we do not multiply by 2 here, this is done Python side!
+        indices.push(tuple.get_item(1)?.extract::<Vec<u32>>()?)
+    }
+
+    let barrier = get_barrier(py, num_qubits as u32);
+
+    let evos = paulis.iter().enumerate().zip(indices).zip(times).flat_map(
+        |(((i, pauli), qubits), time)| {
+            let as_packed = pauli_evolution(pauli, qubits, time, false, do_fountain).map(
+                |(gate, params, qubits)| -> PyResult<Instruction> {
+                    Ok((
+                        gate.into(),
+                        params,
+                        Vec::from_iter(qubits.into_iter()),
+                        Vec::new(),
+                    ))
+                },
+            );
+
+            // this creates an iterator containing a barrier only if required, otherwise it is empty
+            let maybe_barrier = (insert_barriers && i < (num_paulis - 1))
+                .then_some(Ok(barrier.clone()))
+                .into_iter();
+            as_packed.chain(maybe_barrier)
+        },
+    );
+
+    // When handling all-identity Paulis above, we added the time as global phase.
+    // However, the all-identity Paulis should add a negative phase, as they implement
+    // exp(-i t I). We apply the negative sign here, to only do a single (-1) multiplication,
+    // instead of doing it every time we find an all-identity Pauli.
+    if modified_phase {
+        global_phase = multiply_param(&global_phase, -1.0, py);
+    }
+
+    CircuitData::from_packed_operations(py, num_qubits as u32, 0, evos, global_phase)
+}
+
+fn cx_chain(
+    active_paulis: Vec<(char, Qubit)>,
+) -> Box<dyn DoubleEndedIterator<Item = StandardInstruction>> {
+    let num_terms = active_paulis.len();
+    Box::new(
+        (0..num_terms - 1)
+            .map(move |i| (active_paulis[i].1, active_paulis[i + 1].1))
+            .map(|(target, ctrl)| (StandardGate::CXGate, smallvec![], smallvec![ctrl, target])),
+    )
+}
+
+fn cx_fountain(
```

Comment:
could you please add some image to explain how a CX fountain should look like?

Reply:
done in 8d5e3f3